treewide/nixos: remove `with lib;` part 3

Stunkymonkey opened the pull request:
## Description of changes

part of #208242

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

Stunkymonkey committed:
nixos/i18n.inputMethod.hime: remove `with lib;`

Stunkymonkey committed:
nixos/i18n.inputMethod.ibus: remove `with lib;`

Stunkymonkey committed:
nixos/i18n.inputMethod.nabi: remove `with lib;`

Stunkymonkey committed:
nixos/i18n.inputMethod.uim: remove `with lib;`

Stunkymonkey committed:
nixos/label: remove `with lib;`

Stunkymonkey committed:
nixos/crashdump: remove `with lib;`

Stunkymonkey committed:
nixos/assertions: remove `with lib;`

Stunkymonkey committed:
nixos/meta: remove `with lib;`

Stunkymonkey committed:
nixos/fcast-receiver: remove `with lib;`

Stunkymonkey committed:
nixos/security.apparmor: remove `with lib;`

Stunkymonkey committed:
nixos/security.please: remove `with lib;`

Stunkymonkey committed:
nixos/security.audit: remove `with lib;`

Stunkymonkey committed:
nixos/services.oxidized: remove `with lib;`

Stunkymonkey committed:
nixos/services.salt.minion: remove `with lib;`

Stunkymonkey committed:
nixos/services.activemq: remove `with lib;`

Stunkymonkey committed:
nixos/services.gonic: remove `with lib;`

Stunkymonkey committed:
nixos/services.botamusique: remove `with lib;`

Stunkymonkey committed:
nixos/services.gmediarender: remove `with lib;`

Stunkymonkey committed:
nixos/services.goxlr-utility: remove `with lib;`

Stunkymonkey committed:
nixos/services.hqplayerd: remove `with lib;`

Stunkymonkey committed:
nixos/services.jack: remove `with lib;`

Stunkymonkey committed:
nixos/services.icecast: remove `with lib;`

Stunkymonkey committed:
nixos/services.liquidsoap: remove `with lib;`

Stunkymonkey committed:
nixos/services.jmusicbot: remove `with lib;`

Stunkymonkey committed:
nixos/services.roon-bridge: remove `with lib;`

Stunkymonkey committed:
nixos/services.mpdscribble: remove `with lib;`

Stunkymonkey committed:
nixos/services.networkaudiod: remove `with lib;`

Stunkymonkey committed:
nixos/services.spotifyd: remove `with lib;`

Stunkymonkey committed:
nixos/services.roon-server: remove `with lib;`

Stunkymonkey committed:
nixos/services.slimserver: remove `with lib;`

Stunkymonkey committed:
nixos/services.borgmatic: remove `with lib;`

Stunkymonkey committed:
nixos/services.ympd: remove `with lib;`

Stunkymonkey committed:
nixos/services.duplicity: remove `with lib;`

Stunkymonkey committed:
nixos/services.parsoid: remove `with lib;`

Stunkymonkey committed:
nixos/services.duplicati: remove `with lib;`

Stunkymonkey committed:
nixos/services.paperless: remove `with lib;`

Stunkymonkey committed:
nixos/services.plikd: remove `with lib;`

Stunkymonkey committed:
nixos/services.plex: remove `with lib;`

Stunkymonkey committed:
nixos/services.pinnwand: remove `with lib;`

Stunkymonkey committed:
nixos/services.prowlarr: remove `with lib;`

Stunkymonkey committed:
nixos/services.polaris: remove `with lib;`

Stunkymonkey committed:
nixos/services.preload: remove `with lib;`

Stunkymonkey committed:
nixos/services.radarr: remove `with lib;`

Stunkymonkey committed:
nixos/services.pykms: remove `with lib;`

Stunkymonkey committed:
nixos/services.rippleDataApi: remove `with lib;`

Stunkymonkey committed:
nixos/services.readarr: remove `with lib;`

Stunkymonkey committed:
nixos/services.safeeyes: remove `with lib;`

Stunkymonkey committed:
nixos/services.rkvm: remove `with lib;`

Stunkymonkey committed:
nixos/services.rmfakecloud: remove `with lib;`

Stunkymonkey committed:
nixos/services.serviio: remove `with lib;`

Stunkymonkey committed:
nixos/services.sdrplayApi: remove `with lib;`

Stunkymonkey committed:
nixos/services.sickbeard: remove `with lib;`

Stunkymonkey committed:
nixos/services.siproxd: remove `with lib;`

Stunkymonkey committed:
nixos/services.signald: remove `with lib;`

Stunkymonkey committed:
nixos/services.soft-serve: remove `with lib;`

Stunkymonkey committed:
nixos/services.spice-webdavd: remove `with lib;`

Stunkymonkey committed:
nixos/services.sonarr: remove `with lib;`

Stunkymonkey committed:
nixos/services.spice-vdagentd: remove `with lib;`

Stunkymonkey committed:
nixos/services.sundtek: remove `with lib;`

Stunkymonkey committed:
nixos/services.svnserve: remove `with lib;`

Stunkymonkey committed:
nixos/services.sssd: remove `with lib;`

Stunkymonkey committed:
nixos/services.subsonic: remove `with lib;`

Stunkymonkey committed:
nixos/services.synergy: remove `with lib;`

Stunkymonkey committed:
nixos/services.tandoor-recipes: remove `with lib;`

Stunkymonkey committed:
nixos/services.tp-auto-kbbl: remove `with lib;`

Stunkymonkey committed:
nixos/services.tautulli: remove `with lib;`

Stunkymonkey committed:
nixos/services.tiddlywiki: remove `with lib;`

Stunkymonkey committed:
nixos/programs.tuxclocker: remove `with lib;`

Stunkymonkey committed:
nixos/services.uhub: remove `with lib;`

Stunkymonkey committed:
nixos/services.tzupdate: remove `with lib;`

Stunkymonkey committed:
nixos/services.weechat: remove `with lib;`

Stunkymonkey committed:
nixos/services.alerta: remove `with lib;`

Stunkymonkey committed:
nixos/services.zookeeper: remove `with lib;`

Stunkymonkey committed:
nixos/services.xmrig: remove `with lib;`

Stunkymonkey committed:
nixos/services.alloy: remove `with lib;`

Stunkymonkey committed:
nixos/services.apcupsd: remove `with lib;`

Stunkymonkey committed:
nixos/services.arbtt: remove `with lib;`

Stunkymonkey committed:
nixos/services.bosun: remove `with lib;`

Stunkymonkey committed:
nixos/services.cadvisor: remove `with lib;`

Stunkymonkey committed:
nixos/services.below: remove `with lib;`

Stunkymonkey committed:
nixos/services.collectd: remove `with lib;`

Stunkymonkey committed:
nixos/services.das_watchdog: remove `with lib;`

Stunkymonkey committed:
nixos/services.datadog-agent: remove `with lib;`

Stunkymonkey committed:
nixos/services.do-agent: remove `with lib;`

Stunkymonkey committed:
nixos/services.fusionInventory: remove `with lib;`

Stunkymonkey committed:
nixos/services.grafana_reporter: remove `with lib;`

Stunkymonkey committed:
nixos/services.grafana-agent: remove `with lib;`

Stunkymonkey committed:
nixos/services.grafana-image-renderer: remove `with lib;`

Stunkymonkey committed:
nixos/services.karma: remove `with lib;`

Stunkymonkey committed:
nixos/services.kapacitor: remove `with lib;`

Stunkymonkey committed:
nixos/services.incron: remove `with lib;`

Stunkymonkey committed:
nixos/services.hdapsd: remove `with lib;`

Stunkymonkey committed:
nixos/services.heapster: remove `with lib;`

Stunkymonkey committed:
nixos/services.munin-[cron,node]: remove `with lib;`

Stunkymonkey committed:
nixos/services.mackerel-agent: remove `with lib;`

Stunkymonkey committed:
nixos/services.longview: remove `with lib;`

Stunkymonkey committed:
nixos/services.kthxbye: remove `with lib;`

Stunkymonkey committed:
nixos/services.monit: remove `with lib;`

Stunkymonkey committed:
nixos/services.osquery: remove `with lib;`

Stunkymonkey committed:
nixos/services.prometheus.alertmanagerIrcRelay: remove `with lib;`

Stunkymonkey committed:
nixos/services.nagios: remove `with lib;`

Stunkymonkey committed:
nixos/services.netdata: remove `with lib;`

Stunkymonkey committed:
nixos/services.prometheus.pushgateway: remove `with lib;`

Stunkymonkey committed:
nixos/services.prometheus.alertmanagerWebhookLogger: remove `with lib;`

Stunkymonkey committed:
nixos/services.prometheus.alertmanager: remove `with lib;`

Stunkymonkey committed:
nixos/services.prometheus.xmpp-alerts: remove `with lib;`

Stunkymonkey committed:
nixos/services.prometheus.sachet: remove `with lib;`

Stunkymonkey committed:
nixos/services.riemann-tools: remove `with lib;`

Stunkymonkey committed:
nixos/services.riemann-dash: remove `with lib;`

Stunkymonkey committed:
nixos/services.riemann: remove `with lib;`

Stunkymonkey committed:
nixos/services.scollector: remove `with lib;`

Stunkymonkey committed:
nixos/services.smartd: remove `with lib;`

Stunkymonkey committed:
nixos/services.statsd: remove `with lib;`

Stunkymonkey committed:
nixos/services.sysstat: remove `with lib;`

Stunkymonkey committed:
nixos/services.tremor-rs: remove `with lib;`

Stunkymonkey committed:
nixos/services.telegraf: remove `with lib;`

Stunkymonkey committed:
nixos/services.tuptime: remove `with lib;`

Stunkymonkey committed:
nixos/services.uptime-kuma: remove `with lib;`

Stunkymonkey committed:
nixos/services.watchdogd: remove `with lib;`

Stunkymonkey committed:
nixos/services.cachefilesd: remove `with lib;`

Stunkymonkey committed:
nixos/services.vnstat: remove `with lib;`

Stunkymonkey committed:
nixos/services.diod: remove `with lib;`

Stunkymonkey committed:
nixos/services.drbd: remove `with lib;`

Stunkymonkey committed:
nixos/services.nebula: remove `with lib;`

Stunkymonkey committed:
nixos/services.ncdns: remove `with lib;`

Stunkymonkey committed:
nixos/services.glusterfs: remove `with lib;`

Stunkymonkey committed:
nixos/networking.nftables: remove `with lib;`

Stunkymonkey committed:
nixos/services.nylon: remove `with lib;`

Stunkymonkey committed:
nixos/services.oink: remove `with lib;`

Stunkymonkey committed:
nixos/services.stunnel: remove `with lib;`

Stunkymonkey commented:
@philiptaron how should we continue here? do we consider my script as too bad?

philiptaron commented:
> @philiptaron how should we continue here? do we consider my script as too bad?

The fallout has made me pretty reticent to continue. It takes me about an hour to review these, and I'm provably fallible with the defects that have slipped through. I also have to switch over to $work$ for a bit, so I'm a bit slower than I was before.

Ma27 reviewed (dismissed):
AFAIK this is done by some automation: https://github.com/NixOS/nixpkgs/pull/335603#issuecomment-2326936584

Given that this has always caused a notable fallout that maintainers had to fix afterwards, did you fix the script or are we expecting a different result this time?

Stunkymonkey commented:
> AFAIK this is done by some automation: [#335603 (comment)](https://github.com/NixOS/nixpkgs/pull/335603#issuecomment-2326936584)
> 
> Given that this has always caused a notable fallout that maintainers had to fix afterwards, did you fix the script or are we expecting a different result this time?

yes exactly the same script. Automation is not changed.

I was expecting 2 reviewers is enough to detect the problems. Definitely not a fault of the reviewers!

Stunkymonkey commented:
@Ma27 I have implemented a new approach:

using `nix-build nixos/release.nix -A manual.x86_64-linux` to build and parse the output of it with a custom python script. So i am only correcting the reported errors. (based on line-, character-index reports.)

I am doing this until the build succeeds. Yes it is slow, but it surprisingly works:

<details><summary>python-script</summary>

```python
#!/usr/bin/env python3
import subprocess
import re
import sys


def get_option_string():
    result = subprocess.run(
        [
            "nix",
            "eval",
            "--impure",
            "--expr",
            'with import <nixpkgs> {}; builtins.attrNames (builtins.removeAttrs lib ["__unfix__" "removeAttrs" "map" "licenses" "fix" "and" "or" "kernel" "options" "meta" "path"])',
        ],
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
        text=True,
    )
    return result.stdout.replace('" "', " ").replace('" ]', "").replace('[ "', "")


def main():
    print("start post-patch")
    optionstring = get_option_string()
    available_lib_options = set(optionstring.split())
    NIX_BUILD_CMD = "nix-build nixos/release.nix -A manual.x86_64-linux"

    for file_path in sys.argv[1:]:
        while True:
            result = subprocess.run(
                NIX_BUILD_CMD.split(),
                stdout=subprocess.PIPE,
                stderr=subprocess.PIPE,
                text=True,
            )
            error_output = result.stderr
            reversed_error = error_output.splitlines()[::-1]
            error_variable_match = re.search(
                r"undefined variable '(\w+)'", error_output
            )
            # print(error_output, error_variable_match)
            if error_variable_match:
                error_variable = error_variable_match.group(1)
                line_info = next(
                    (line for line in reversed_error if "  at " in line), None
                )
                if line_info and error_variable in available_lib_options:
                    print(f"loc: {line_info.strip()}")
                    line_number = (
                        int(re.search(r":([0-9]+):[0-9]+:", line_info).group(1)) - 1
                    )
                    char_number = (
                        int(re.search(r":[0-9]+:([0-9]+):", line_info).group(1)) - 1
                    )
                    with open(file_path, "r") as file:
                        lines = file.readlines()
                    error_line = lines[line_number]
                    print(f"old: {error_line.strip()}")
                    modified_line = (
                        error_line[:char_number] + "lib." + error_line[char_number:]
                    )
                    print(f"new: {modified_line.strip()}")
                    lines[line_number] = modified_line
                    with open(file_path, "w") as file:
                        file.writelines(lines)
            else:
                break


if __name__ == "__main__":
    main()
```

</details>

is this a better approach? are you fine with it?

Ma27 commented:
> is this a better approach? are you fine with it?

I think it's a better approach than what we had before.
Another idea: evaluate relevant VM tests (if any): this approach covers option declarations and stuff around those, but I think e.g. code that puts systemd untis together isn't affected?
It may also help to do `<module>.enable = true;`, evaluate and check if something happens that isn't a ("X is used but not defined")?

That being said, my main concern was that seemingly there didn't happen much QA during the past iterations and we somehow expected different results on followups, hence my first comment.

But, after going through the options, I don't really have a good answer on how to test these kinds of changes.
@Infinisil or @roberth you perhaps?

Stunkymonkey commented:
@Ma27 should we split each commit into a separate MR? this way it would take longer, but the separation is better.

philiptaron commented:
I'm going to give it a human review and see the error rate that I can find.

philiptaron reviewed (approved):
I can't find a single error of commission in this PR, meaning that (a) the error rate for this new script is much lower and (b) whatever errors this script produces are going to be errors of omission rather than commission.

I'm willing to merge it given the review.

In particular, given this line in the script:

```
with import <nixpkgs> {};
builtins.attrNames (builtins.removeAttrs lib ["__unfix__" "removeAttrs" "map" "licenses" "fix" "and" "or" "kernel" "options" "meta" "path"])
```

The likely errors are places where `licenses`, `fix`, `and`, `or`, `kernel`, `options`, `meta`, or `path` are actually used in a NixOS module, as the script will skip them. `lib.or` is such a strange beast that I am going to discount that one.

Any file which does not contain these tokens will not have an error, of that I am fairly certain. If a file _does_ contain these tokens, the context needs to be taken into account -- most of the uses are unlikely to be errors.

Stunkymonkey commented on nixos/modules/services/amqp/activemq/default.nix:
the best part about it: even less work for me...

will rebase and start working on the next part

philiptaron commented on nixos/modules/services/misc/paperless.nix:
```diff
@@ -229,29 +227,11 @@ in
 
       This sets `OMP_NUM_THREADS` to `1` in order to mitigate the issue. See
       https://github.com/NixOS/nixpkgs/issues/240591 for more information
-    '' // mkOption { default = true; };
```
this isn't expected, right, @Stunkymonkey?

Stunkymonkey replied:
you are totally right, might be i did something wrong while rebasing.

I reverted all my changes and rerun the script.

philiptaron replied:
On this file only, I hope!

philiptaron commented on nixos/modules/services/misc/redlib.nix:
```diff
@@ -88,8 +87,7 @@ in
   };
 
-  config = mkIf cfg.enable {
-    systemd.packages = [ cfg.package ];
```
Here's another rebase error, I'm assuming.

Stunkymonkey replied:
removed the commit touching that file. already resolved via e286b91ebca6f7de81cefb87492440582a565990

philiptaron commented:
Let's ship it and see what shakes out.

philiptaron commented:
2 days later and no error reports, just a rebase ping. Way to go Felix!
